File: sources/kotlin/15_file_write.kt
// task 15 file_write — expected output: 104857600
// build: kotlinc 15_file_write.kt -include-runtime -d prog.jar    run: java -jar prog.jar    (jvm row; the native row builds 15_file_write_native.kt)
// note: java.io is not available on Kotlin/Native, so the native row has its own file, which
// writes out.bin with platform.posix instead. This one writes it in 1 MiB chunks.

import java.io.FileOutputStream

fun main() {
    val chunk = 1024 * 1024
    val buf = ByteArray(chunk)
    for (i in 0 until chunk) {
        buf[i] = (i % 256).toByte()
    }
    var written = 0L
    FileOutputStream("out.bin").use { out ->
        for (t in 0 until 100) {
            out.write(buf)
            written += chunk.toLong()
        }
        out.flush()
        out.getFD().sync()
    }
    println(written)
}

File: sources/rust/15_file_write.rs
// task 15 file_write — expected output: 104857600
// build: rustc -O -o prog 15_file_write.rs    run: ./prog

use std::fs::File;
use std::io::Write;

fn main() {
    let mut buf = vec![0u8; 1024 * 1024];
    for i in 0..buf.len() {
        buf[i] = (i % 256) as u8;
    }

    let mut file = File::create("out.bin").expect("cannot create out.bin");
    let mut written: u64 = 0;

    for _ in 0..100 {
        file.write_all(&buf).expect("write failed");
        written += buf.len() as u64;
    }

    file.flush().expect("flush failed");
    file.sync_all().expect("sync failed");

    println!("{}", written);
}
